### ISSUE TITLE
Add PGO support

### ISSUE THREAD

**Issue by zamazan4ik** [RustPython/RustPython#5000]
## Summary

Add the possibility to compile RustPython with Profile-Guided Optimization (PGO) support to improve the performance even further.

## Details

CPython [has](https://github.com/python/cpython#profile-guided-optimization) PGO support. Benefits from PGO for CPython are described [here](https://www.activestate.com/blog/python-performance-boost-using-profile-guided-optimization/). I think a similar approach could be used for RustPython as well.

Some materials about PGO (implementation details, caveats, project list with PGO, etc) could be found here: https://github.com/ZaMaZaN4iK/awesome-pgo

**Comment by DimitrisJim** [comment]
would you be looking for a script file to build the binary with pgo? I don't think writting it would be too difficult. Can't speak for how long the build would take in this case, though.

**Comment by zamazan4ik** [comment]
> would you be looking for a script file to build the binary with pgo?

It's up to you how to implement PGO integration into the project. Like it's already done in CPython (yes, a part of the build scripts), maybe somehow else (probably by using https://github.com/Kobzol/cargo-pgo). The main question here is the sample load. Not sure, which load is used by CPython.

**Comment by DimitrisJim** [comment]
> The main question here is the sample load

Yup. Same as CPython probably, subset of test-suite with `--pgo` (the references you found) or a bigger load with `--pgo-extended` iirc. Sounds like something that would be nice to have and probably best addressed via `cargo-pgo` steps in some README.

**Comment by DimitrisJim** [comment]
Note: I thought about this and a wiki entry seems the right place for it. Didn't get to running the steps yet since I just created the entry yesterday but feel free to keep an eye on it https://github.com/RustPython/RustPython/wiki/Build-an-instrumented-binary-with-cargo-pgo

I'll update this comment when I'm done with it.
